Azure REST API Guidelines Update: RC1 (#264)
Microsoft/api-guidelines: azure/Guidelines.md

#### Exactly Once Behavior = Client Retries & Service Idempotency
> :white_check_mark: **DO** ensure that __all__ HTTP methods are idempotent.
 
> :ballot_box_with_check: **YOU SHOULD** use PUT or PATCH to create a resource as these HTTP methods are easy to implement, allow the customer to name their own resource, and are idempotent.
 
> :heavy_check_mark: **YOU MAY** use POST to create a resource but you must make it idempotent and, of course, the response __must__ return the URL of the created resource with a 201-Created. One way to make POST idempotent is to use the Repeatability-Request-ID & Repeatability-First-Sent headers (See [Repeatability of requests](#Repeatability-of-requests)).

##### HTTP Return Codes
> :white_check_mark: **DO** adhere to the return codes in the following table when the method is successful:

Method | Description | Response Status Code 
-------|-------------|---------------------
PATCH  | Create/Modify the resource with JSON Merge Patch | 200-OK, 201-Created

> PUT    | Create/Replace the *whole* resource | 200-OK, 201-Created 

Reviewer: No 202 LRO? Genuine question, I'm so used to see that, that I don't know anymore if it's actually good or bad :p

Author: Yes on supporting 202 on PUT/POST/DELETE.

In the LRO section, we will add additional guidance on these.

> :white_check_mark: **DO** validate all query parameter and request header values and fail the operation with ```400-Bad Request``` if any value fails validation. Return an error response as described in the [Handling Errors](#Handling-errors) section indicating what is wrong so customer can diagnose the issue and fix it themselves.
> 
> :white_check_mark: **DO** use the following table when translating strings:

Data type | Document that string must be
--------- | -------
Boolean   | true / false (all lowercase)
Integer   | -2<sup>53</sup>+1 to +2<sup>53</sup>-1 (for consistency with JSON limits on integers [RFC8259](https://datatracker.ietf.org/doc/html/rfc8259))
Float     | [IEEE-754 binary64](https://en.wikipedia.org/wiki/Double-precision_floating-point_format)
String    | (Un)quoted?, max length, legal characters, case-sensitive, multiple delimiter
UUID      | 123e4567-e89b-12d3-a456-426614174000 (no {}s, hyphens, case-insensitive) [RFC4122](https://datatracker.ietf.org/doc/html/rfc4122)
Date/Time (Header) | Sun, 06 Nov 1994 08:49:37 GMT [RFC7231](https://datatracker.ietf.org/doc/html/rfc7231#section-7.1.1.1)

> Date/Time (Query parameter) | YYYY-MM-DDTHH:mm:ss.sssZ [RFC3339](https://datatracker.ietf.org/doc/html/rfc3339)

Reviewer: I would do a parenthesis like above to rewrite one more time "3 digits" so people can't miss it. Got too much hurt by this :)

Author: @lmazuel -- not sure what you are referring to on the parenthesis. however, we think you want us to emphasize that the time MUST be to three digits. We'll take a pass at this.

> :ballot_box_with_check: **YOU SHOULD** support paging today if there is ever a chance in the future that the number of items can grow to be very large. 
> NOTE: It is a breaking change to add paging in the future

> :heavy_check_mark: **YOU MAY** expose an operation that lists your resources by supporting a GET method with a URL to a resource-collection (as opposed to a resource-id).

**Example Response Body** 
```
{
    "value": [
       { "id": "Item 01", "etag": "0xabc", "price": 99.95, "sizes": null },
       { … },
       { … },
       { "id": "Item 99", "etag": "0xdef", "price": 59.99, "sizes": null }
    ],
    "nextLink": "{opaqueUrl}"
 }
```

> :white_check_mark: **DO** include the id field and etag field (if supported) for each item as this allows the customer to modify the item in a future operation.
> 
> :white_check_mark: **DO** clearly document that resources may be skipped or duplicated across pages of a paginated collection unless the operation has made special provisions to prevent this (like taking a time-expiring snapshot of the collection).
>

> > :white_check_mark: **DO** return a ```nextLink``` field with a URL that the client can GET in order to retrieve the next page of the collection.

Reviewer: Should we clarify it should be an absolute URL callable as-is? We have precedence of people expecting people and/or SDK to rebuilt the full url with some host configuration, which has been painful in the past.

Author: +1. Will update

> :ballot_box_with_check: **YOU SHOULD** use ```value``` as the name of the top-level array field unless a more appropriate name is available.

> > :no_entry: **DO NOT** return the ```nextLink``` field at all when returning the last page of the collection. 

Reviewer: Should we make it explit, when we suggest people to support this format even if they are not pageable yet, that it means their only page is implictly "the last page", and therefore it's mostly modeling the API like that (IOW, with "value" but no "nextLink")?

Author: @lmazuel -- The way we're interpreting this is... "If you only return one page, it's implicitly your last page." In this case, there would be no "nextLink" field. Is what you are suggesting that services should document their API "as if" it was always "pageable"?

If we got the gist right, would you mind suggesting some text as a starting point for us?

Author: Decision: We think this is sufficient at this point.

Reviewer: Nope, your interpretation is what I meant, sorry if I wasn't clear :)

Reviewer 2: We know of several listing operations that are not now and never will be pageable. For example, returning a set of countries or states. Because the client code to process a collection (loop for each page) is inconvenient, we do NOT want to force all list operations to support this pattern which would then require that all customer code write loops.

>
> :no_entry: **DO NOT** ever return a ```nextLink``` field with a value of null.

#### Query options
> :heavy_check_mark: **YOU MAY** support the following query parameters allowing customers to control the list operation:

Parameter&nbsp;name | Type | Description
------------------- | ---- | -----------
_filter_       | string            | an expression on the resource type that selects the resources to be returned

> _orderby_      | string&nbsp;array | a list of expressions that specify the order of the returned resources

Reviewer: Do we have opinions/guidance if service team should support expanded array, group array by comma, or both?

IOW: `a=b&a=c` VS `a=b,c`

The comma joined version has interesting impact on encoding, especially if a field in a DB contains a comma: `query=%2C,%2C,%2C`

Author: +1. We will add a SHOULD for supporting commas. If user defined input, then escape the comma.

Reviewer 3: I updated the "HTTP Query Parameters and Header Values" section, where it describes how query params should be represented as strings, to cover arrays.  I said two formats are allowed, csv (preferred) and multi (separate name=value for each array item).  We can adjust there if want additional or different formats.

Author: +1

A special case of such an expression is a property path terminating on a primitive property.

Each expression in the _orderby_ parameter value may include the suffix "asc" for ascending or "desc" for descending, separated from the expression by one or more spaces.

> :white_check_mark: **DO** sort the collection in ascending order on an expression if "asc" or "desc" is not specified.
> 
> :white_check_mark: **DO** sort NULL values as "less than" non-NULL values.
> 
> :white_check_mark: **DO** sort items by the result values of the first expression, and then sort items with the same value for the first expression by the result value of the second expression, and so on.
> 
> :white_check_mark: **DO** use the inherent sort order for the type of the field. For example, date-time values should be sorted chronologically and not alphabetically.
> 
> :white_check_mark: **DO** respond with an error message as defined in the [Unsupported Requests](??) section if the client requests sorting by a field that is not supported by the operation.

For example, to return all people sorted by name in ascending order:
```http

> GET https://api.contoso.com/people?$orderBy=name

Reviewer: It's said earlier no prefix with $

Author: Good catch. Thank you.

```

For example, to return all people sorted by name in descending order and a secondary sort order of hireDate in ascending order.
```http

> GET https://api.contoso.com/people?$orderBy=name desc,hireDate

Reviewer: It's said earlier no prefix with $

Author: Good catch. Thank you.

If supporting _maxpagesize_

> :white_check_mark: **DO** define the _maxpagesize_ parameter as an optional integer with a default value appropriate for the collection.
> 
> :white_check_mark: **DO** make clear in documentation of the _maxpagesize_ parameter that the operation may choose to return fewer resources than the value specified.

### API Versioning

Azure services need to change over time. However, when changing a service, there are 2 requirements:
 1. Already-running customer workloads must never break due to a service change
 2. Customers can adopt a new service version without requiring any code changes (Of course, the customer must modify code to leverage any new service features.)

> :white_check_mark: **DO** review any API changes with the Azure API Stewardship Board
> 
> :white_check_mark: **DO** use an 'api-version' query parameter with a date value
> ```text PUT https://service.azure.com/users/Jeff?api-version=2021-06-04```
>
> :white_check_mark: **DO** use a later date for each new preview version
> When releasing a new preview, the service team may completely retire any previous preview versions after giving customers at least 90 days to upgrade their code

> > :white_check_mark: **DO** use a later date for successive preview versions. 

Reviewer: Miss a newline above

Author: Good catch. Thank you.

> :ballot_box_with_check: **YOU SHOULD** support repeatable requests according as defined in [OASIS Repeatable Requests Version 1.0](https://docs.oasis-open.org/odata/repeatable-requests/v1.0/repeatable-requests-v1.0.html).

- The tracked time window (difference between the `Repeatability-First-Sent` value and the current time) **MUST** be at least 5 minutes.
- A service advertises support for repeatability requests by adding the `Repeatability-First-Sent` and `Repeatability-Request-ID` to the set of headers for a given operation.
- When understood, all endpoints co-located behind a DNS name **MUST** understand the header. This means that a service **MUST NOT** ignore the presence of a header for any endpoints behind the DNS name, but rather fail the request containing a `Repeatability-Request-ID` header if that particular endpoint lacks support for repeatable requests. Such partial support **SHOULD** be avoided due to the confusion it causes for clients.

<!-- Open API Spec -->
[OpenAPI Specification]: https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md

### Long Running Operations & Jobs

> The Microsoft REST API guidelines for Long Running Operations are an updated, clarified and simplified version of the Asynchronous Operations guidelines from the 2.1 version of the Azure API guidelines. Unfortunately, to generalize to the whole of Microsoft and not just Azure, the HEADER used in the operation was renamed from `Azure-AsyncOperation` to `Operation-Location`. 

Reviewer: Are we planning another document on details of this? Like Location headers, the format of Operation-Location, Resource-Location, etc.

Author: At this point, we were not planning a separate document on these details. 
We're inferring from your comment that we need more guidance on LROs. If that's the case, let's open an issue as a placeholder so the we can discuss the right way to capture this information.

Author: Here's the link to the issue that we'll use to revisit LROs.

##### Single file access
Desiging an API for accessing a single file, depending on your scenario, is relatively straight forward.

> :heavy_check_mark: **YOU MAY** use a Shared Access Signature [SAS](https://docs.microsoft.com/en-us/azure/storage/common/storage-sas-overview) to provide access to a single file. SAS is considered the minimum security for files and can be used in lieu of, or in addition to, RBAC.   

> :ballot_box_with_check: **YOU SHOULD** if using HTTP (not HTTPS) document to users that all information is sent over the wire in clear text.  
> 
> :ballot_box_with_check: **YOU SHOULD** support managed identity using Azure Storage by default (if using Azure services). 

###### File Versioning
Depending on your requirements, there are scenarios where users of your service will require a specific version of a file. For example, you may need to keep track of configuration changes over time to be able to rollback to a previous state. In these scenarios, you will need to provide a mechanism for accessing a specific version.

> :white_check_mark: **DO** Enable the customer to provide an ETag to specify a specific version of a file. 
##### File Collections

> When your users need to work with multiple files, for example a document translation service, it will be important to provide them access to the collection, and its contents, in a consistent manner. Because there is no industry standard for working with with containers, these guidelines will recommend that you leverage Azure Storage. Following the guidelines above, you also want to ensure that you don't expose file system constructs, e.g. folders, and instead use storage constructs, e.g. blob prefixes.

Reviewer: ```suggestion
When your users need to work with multiple files, for example a document translation service, it will be important to provide them access to the collection, and its contents, in a consistent manner. Because there is no industry standard for working with containers, these guidelines will recommend that you leverage Azure Storage. Following the guidelines above, you also want to ensure that you don't expose file system constructs, e.g. folders, and instead use storage constructs, e.g. blob prefixes.
```

Author: Good catch. Thank you.

| PATCH / PUT | if-none-match | *     | check for *any* version of the resource, if one is found, fail the operation |  412 Precondition Failed | Response body SHOULD return the serialized value of the resource (typically JSON) that was passed along with the request.|
| PATCH / PUT | if-match | value of etag     | value of if-match equals the latest etag value on the server, confirming that the version of the resource is the most current | 200 OK or </br> 201 Created </br> | Response header MUST include the new ```ETag``` value. Response body SHOULD include the serialized value of the resource (typically JSON).  |
| PATCH / PUT | if-match | value of etag     | value of if-match header DOES NOT equal the latest etag value on the server, indicating a change has ocurred since after the client fetched the resource|  412 Precondition Failed | Response body SHOULD return the serialized value of the resource (typically JSON) that was passed along with the request.|

> | DELETE      | if-none-match | value of etag     | value does NOT match the latest value on the server | 412 Preconditioned Failed | Response body SHOULD return the serialized value of the resource (typically JSON) that was passed along with the request.  |

Author: ```suggestion
| DELETE      | if-none-match | value of etag     | value does NOT match the latest value on the server | 412 Preconditioned Failed | Response body SHOULD be empty.|
```

> | DELETE      | if-none-match | value of etag     | value matches the latest value on the server | 200 OK or </br> 204 No Content | Response body SHOULD return the serialized value of the resource (typically JSON) that was passed along with the request.  |

Author: ```suggestion
| DELETE      | if-none-match | value of etag     | value matches the latest value on the server | 200 OK or </br> 204 No Content | Response body SHOULD be empty.  |
```

POST   | Create new resource (ID set by service) | 201-Created with URL of created resource
GET    | Read (i.e. list) a resource collection | 200-OK
GET    | Read the resource | 200-OK 
DELETE | Remove the resource | 204-No Content\; avoid 404-Not Found 

> :white_check_mark: **DO** return status code ```202-Accepted``` and follow the guidance in [Long Running Operations & Jobs](#long-running-operations--jobs) when a PUT, POST, or DELETE method completes asynchronously

> :white_check_mark: **DO** treat method names as case sensitive and should always be in uppercase
> 
> :white_check_mark: **DO** return the state of the resource after a PUT, PATCH, POST, or GET operation with a ```200-OK``` or ```201-Created```.
> 
> :white_check_mark: **DO** return a ```204-No Content``` without a resource/body for a DELETE operation (even if the URL identifies a resource that does not exist; do not return ```404-Not Found```)
> 
> :white_check_mark: **DO** support caching and optimistic concurrency by honoring the the if-match, if-none-match, if-modified-since, and if-unmodified-since request headers and by returning the etag and last-modified response headers

> 

Reviewer 4: Consider adding:

“DO return a 404 when the user does not have access to the resource.  Returning a 404, vs. a 403, prevents a malicious request from learning anything about resources based on error codes.”

Reviewer 3: Added.

Reviewer 5: the safety of this additional recommendation isn't clear to me. Doesn't returning 404 on "right resource, no access" similarly enable a fuzzer to learn something?" isn't 403 beneficial here too?

Reviewer 6: I believe ARM returns a 403 in this case at present across all RPs in the management plane if the caller does not have permissions to a scope regardless of whether that scope exists or not.  I had to investigate a customer incident just last week where this behavior was observed.

Reviewer 7: Perhaps the wording was unclear, but the intent is to return a 404 _whether or not the resource exists_ if the caller does not have the appropriate permissions.  In other words, check permission first, but if permission fails that return 404.

Reviewer 5: understand now. I recommend we follow the ARM convention as per Michael. The Azure Notebooks Resource implementation (my team's stuff) follows the "always return 403" guidance per the control plane docs. I think this offers better security through obscurity.

Reviewer 8: I think in the data plane the 404 for permission denied is more appropriate, even if it is different than the ARM guidance.  In ARM it makes sense that the tooling should differentiate because that information is likely useful and/or important for tooling or automation built on the ARM APIs.  For the data plane we have *many* more use cases and in many of them it makes sense to not reveal information in many of those cases.

Reviewer 8: to expand - the resource to which you don't have permission in the data plane is likely to be content that belongs to a tenant organization, whereas in the control plane it is just a cloud resource.  Revealing the existence (or not) of particular pieces of content/data is more sensitive in those cases.

Reviewer 6: I think there is some confusion here -- ARM does not return 403s only when a resource exists but the user is not authorized.  It always returns 403s when a user doesn't have access to the scope (regardless of whether or not that scope exists or not).  The key requirement here is to not disclose additional details to an unauthorized user -- being consistent accomplishes that regardless of whether or not you consistently return a 403 or a 404.  Returning a 403 is consistent with current ARM semantics which RPs already need to account for both in a producing and consumption role.

Author: We updated the guidance to be in alignment with ARM. 
Resolving this thread.

> # Microsoft Azure REST API Guidelines

Reviewer 5: Are websockets and negotiating their creation out of scope of this document, or indeed in scope for the larger standardization effort in general? Several services in DevDiv use these and i think guidance would be valuable.

Author: It's out of scope for this document, but we want to add a separate doc for this document.
 https://github.com/microsoft/api-guidelines/issues/272

## History

| Date        | Version | Notes                                               |
| ----------- | ------- | --------------------------------------------------- |
| 2021-Jul-02 | vNext-RC1 | Updated Azure REST Guidelines.                    |
| 2020-Jul-31 | v3.2    | Added service advice for initial versions           |
| 2020-Mar-31 | v3.1    | 1st public release of the Azure REST API Guidelines |

## Introduction

These guidelines offer prescriptive guidance that Azure service teams MUST follow ensuring that customers have a great experience by designing APIs meeting these goals:
- Developer friendly via consistent patterns & web standards (HTTP, REST, JSON)
- Efficient & cost-effective
- Work well with SDKs in many programming languages

> - Customers can create fault-tolerant apps by supporting retries/idempotency/optimistic concurrency

Reviewer 6: There is very little documentation in terms of guidance to service owners on how to properly implement idempotency and that has led to unexpected behavior from RP to RP -- e.g. my team had an issue related to an idempotency implementation by a downstream RP where they used a single header value to uniquely identify requests (regardless of content) and return cached responses without doing any further work or returning an error.  Is this guidance coming?  The payment industry has some pretty solid semantics around idempotency that work quite well.  Stripe had a pretty extensive backend around supporting idempotency while also informing customers when idempotency semantics were being violated.

Author: @michaelwstark --
We are going to add content and/or a section to the "considerations" document. We'll also look to add a sentence or two here for clarity.

Reviewer 3: I've added a brief discussion of idempotency to the Considerations doc -- starting at line 94 -- which also points into the detailed guidelines in the Guidelines doc.

Reviewer 9: There is ongoing work in the IETF httpapi working group to standardized the Idempotency-Key header used by Stripe. https://github.com/ietf-wg-httpapi/idempotency

Author: @michaelwstark -- We've taken the action to reconcile this with the RPs ( see: #273 ). 
@darrelmiller -- we are going to review the Stripe doc and determine if this or the Oasis guidance is what we should include. However, we will do this as a distinct work item (see: #274 )

The Microsoft Azure Cloud platform exposes its APIs through the core building blocks of the Internet; namely HTTP, REST, and JSON. This section provides you with a general understanding of how these technologies should be applied when creating your service.

### HTTP
Azure services must adhere to the HTTP specification, [RFC7231](https://tools.ietf.org/html/rfc7231). This section further refines and constrains how service implementors should apply the constructs defined in the HTTP specification. It is therefore, important that you have a firm understanding of the following concepts:
* [Uniform Resource Locators (URLs)](URLS)
* HTTP Methods
* Request & Response Headers
* Bodies

### Uniform Resource Locators (URLs)

A Uniform Resource Locator (URL) is how developers access the resources of your service. Ultimately, URLs are how developers form a cognitive model of your service's resources.

> :white_check_mark: **DO** use this URL pattern:
```text
https://<service>.<cloud>/<tenant>/<service-root>/<resource-collection>/<resource-id>/
```

Where:
 | Field | Description
 | - | - |
 | service | Name of the service (ex: blobstore, servicebus, directory, or management)

>  | cloud | Cloud domain name (see Azure CLI's "az cloud list")<p><table><tr><td><b>Cloud</td><td><b>Domain</td></tr><tr><td>Public</td><td>azure.net</td></tr><tr><td>US Government</td><td>usgovcloudapi.net</td></tr><tr><td>China</td><td>chinacloudapi.cn</td></tr><tr><td>German</td><td>cloudapi.de</td></tr></table>

Reviewer 5: perhaps note that black forest is being deprecated. 
we should also include Fairfax and Mooncake aliases for usgov and china

 | tenant | Globally-unique ID of container representing tenant isolation, billing, enforced quotas, lifetime of containees (ex: subscription UUID)
 | service&#x2011;root | Service-specific path (ex: blobcontainer, myqueue)
 | resource&#x2011;collection | Name of the collection, unabbreviated, pluralized
 | resource&#x2011;id | Value of the unique id property. This MUST be the raw string/number/guid value with no quoting but properly escaped to fit in a URL segment.

> :white_check_mark: **DO** use kebab-casing (preferred) or camel-casing for URL path segments. If the segment refers to a JSON field, use camel casing.
> 
> :white_check_mark: **DO** return ```414-URI Too Long``` if a URL exceeds 2083 characters
>

> > :white_check_mark: **DO** treat URL path segments as case-sensitive. If the passed-in case doesn't match what the service expects, the request __MUST__ fail with the appropriate HTTP return code.

Reviewer 10: What are the appropriate codes for this condition?

Author: ```suggestion
> :white_check_mark: **DO** treat URL path segments as case-sensitive. If the passed-in case doesn't match what the service expects, the request __MUST__ fail with the HTTP ```404 Not Found``` return code.
```

#### Exactly Once Behavior = Client Retries & Service Idempotency
> :white_check_mark: **DO** ensure that __all__ HTTP methods are idempotent.
 
> :ballot_box_with_check: **YOU SHOULD** use PUT or PATCH to create a resource as these HTTP methods are easy to implement, allow the customer to name their own resource, and are idempotent.
 
> :heavy_check_mark: **YOU MAY** use POST to create a resource but you must make it idempotent and, of course, the response __must__ return the URL of the created resource with a 201-Created. One way to make POST idempotent is to use the Repeatability-Request-ID & Repeatability-First-Sent headers (See [Repeatability of requests](#Repeatability-of-requests)).

##### HTTP Return Codes
> :white_check_mark: **DO** adhere to the return codes in the following table when the method completes synchronously and is successful:

> Method | Description | Response Status Code 

Reviewer 5: The Azure ARM team maintains a API review process and linter for Resource Provider APIs. Do we plan to review the guidance in this doc with RP guidelines so we can be sure where we are necessarily different? Consistent guidance across control and data plane will be easier to enforce and learn, with deviations where appropriate.

Reviewer 7: We should make sure these guidelines are consistent with the RP guidelines — meaning that these guidelines should not contradict anything in the RP guidelines, but the RP guidelines can be more strict.

Reviewer 5: thanks mike, is that happening, or do you need some introductions to folks?

Reviewer 3: I've opened issue #273 to track this work.

> 
> :white_check_mark: **DO** return the state of the resource after a PUT, PATCH, POST, or GET operation with a ```200-OK``` or ```201-Created```.
> 
> :white_check_mark: **DO** return a ```204-No Content``` without a resource/body for a DELETE operation (even if the URL identifies a resource that does not exist; do not return ```404-Not Found```)
> 
> :white_check_mark: **DO** support caching and optimistic concurrency by honoring the the if-match, if-none-match, if-modified-since, and if-unmodified-since request headers and by returning the etag and last-modified response headers

### HTTP Query Parameters and Header Values
Because information in the service URL, as well as the request / response, are strings, there must be a predictable, well-defined scheme to convert strings to their corresponding values.

> :white_check_mark: **DO** validate all query parameter and request header values and fail the operation with ```400-Bad Request``` if any value fails validation. Return an error response as described in the [Handling Errors](#Handling-errors) section indicating what is wrong so customer can diagnose the issue and fix it themselves.
> 

> > :white_check_mark: **DO** use the following table when translating strings:

Reviewer 10: It seems like Azure services are often implemented in .NET. I wonder if we should not provide a .NET library that would help service teams implement some of the behaviors recommended by this document. Possibly some of these APIs already exist in .NET, and in such case, we would just provide a sample/guide how to use them, not a library.

Author: Great idea!

* [Why isn't HTTP PUT allowed to do partial updates in a REST API?](https://stackoverflow.com/questions/19732423/why-isnt-http-put-allowed-to-do-partial-updates-in-a-rest-api)

### REpresentational State Transfer (REST)
REST is an architectural style with broad reach that emphasizes scalability, generality, independent deployment, reduced latency via caching, and security. When applying REST to your API, you define your service’s resources as a collections of items. These are typically the nouns you use in the vocabulary of your service. Your service's [URLs](#URLS) determine the hierarchical path developers use to perform CRUD (create, read, update, and delete) operations on resources. Note, it's important to model resource state, not behavior. There are patterns, later in these guidelines, that describe how to invoke behavior on your service. See [this article in the Azure Architecture Center](https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design) for a more detailed discussion of REST API design patterns.

When designing your service, it is important to optimize for the developer using your API.

> > :white_check_mark: **DO** focus heavily on clear & consistent naming

Reviewer 10: What is "clear" naming? :-) I know that a book could be written on this, but maybe we could give a couple of tips to service teams: 1) don't invent fancy terms or use fancy words. Explain the concept you want to name to an average dev who is not a domain expert and then name the concept using similar verbage, 2) Don't use too generic names. Names should highlight how this concept is different from other similar concepts the users is likely to run into when they program Azure. 3) Pick one word/term out of a set of synonyms and stick to it, i.e. don't use synonyms; you are not writing a literature paper.

Author: +1. Will add some clarity

Reviewer 3: Guidance added to [`ConsiderationsForServiceDesign.md`](https://github.com/microsoft/api-guidelines/pull/264/files#diff-4a9da7ae492c4eab6b49f0bfb120cd25b6b1ae49f647bc74ecf04a48d8fe2752R24)

Author: +1. LGTM.

#### Enums & SDKs (Client libraries)
It is common for strings to have an explicit set of values. These are often reflected in the OpenAPI definition as enumerations. These are extremely useful for developer tooling, e.g. code completion, and client library generation.

However, it is not uncommon for the set of values to grow over the life of a service. For this reason, Microsoft's tooling uses the concept of an "extensible enum," which indicates that the set of values should be treated as only a *partial* list.
This indicates to client libraries and customers that values of the enumeration field should be effectively treated as strings and that undocumented value may returned in the future. This enables the set of values to grow over time while ensuring stability in client libraries and customer code.

> :ballot_box_with_check: **YOU SHOULD** use extensible enumerations unless you are positive that the symbol set will NEVER change over time.
> 
> :white_check_mark: **DO** document to customers that new values may appear in the future so that customers write their code today expecting these new values tomorrow.
 

> > :no_entry: **DO NOT** send "enum integers" over the wire.

Reviewer 10: It might be good to add justification for such strong guideline.

Author: +1. Will add some clarity.

Reviewer 3: I am going to remove this guidance and opened #271 to document the problem and drive resolution.

> 
> :no_entry: **DO NOT** remove values from your enumeration list as this breaks customer code.

#### Polymorphic types
> :warning: **YOU SHOULD NOT** use polymorphic JSON types because they greatly complicate the customer code due to runtime dynamic casts and the introduction of new types in the future. 

If you can't avoid them, then follow the guideline below.

> > :white_check_mark: **DO** define a ```kind``` field indicating the kind of the resource and include any kind-specific fields in the body. 

Reviewer 10: Can we mandate that this filed must be first? Also, I am not sure this guideline is enough to be able to efficiently parse such JSON in .NET. I think it would be easier/faster if the kind property was outside of the object data, i.e. { kind : "...', value : ... }

Author: Including "kind" in this structure is the way that OpenAPI specifies how you do this and it will lead to problems. Because JSON is inherently unordered, and we can't guarantee how this will be serialized, we should not do anything that requires an order.

Reviewer 2: This pattern is proper for JSON objects. It may be more difficult for .NET (and other languages) to parse but we also need to be true to REST/JSON customers.

> :heavy_check_mark: **YOU MAY** allow clients to pass the _skip_ query parameter to specify an offset into collection of the first resource to be returned.
##### top

> :heavy_check_mark: **YOU MAY** allow clients to pass the _top_ query parameter to specify the maximum number of resources to return from the collection.

If supporting _top_:
> :white_check_mark: **DO** define the _top_ parameter as an integer with a minimum value of 1. If not specified, __top__ has a default value of infinity.
> 
> :white_check_mark: **DO** return the collection's _top_ number of resources (if available), starting from _skip_.

##### maxpagesize

> :heavy_check_mark: **YOU MAY** allow clients to pass the _maxpagesize_ query parameter to specify the maximum number of resources to include in a single page response.

If supporting _maxpagesize_

> :white_check_mark: **DO** define the _maxpagesize_ parameter as an optional integer with a default value appropriate for the collection.
> 
> :white_check_mark: **DO** make clear in documentation of the _maxpagesize_ parameter that the operation may choose to return fewer resources than the value specified.

> ### API Versioning

Reviewer 5: OSS
- Semantic versioning is popular in OSS, do we have an opinion on that pattern in this doc? 
- Some Microsoft services proxy calls to OSS Rest APIs. These will have their own semantics. Are there any considerations for how to frame/expose these within an API following MS conventions?

Reviewer 4: On semantic versioning, we're requiring services to always be backwards compatible which obviates a lot of the need for semver.  We prefer date based versioning as outlined below and in https://www.youtube.com/watch?v=bG984_L6dS0&list=PL9XzOCngAkqs4m0XdULJu_78nM3Ok3Q65&index=14.

Reviewer 3: If a service exposes an "industry standard" API of OSS software for compatibility / interoperability reasons, that may justify some exceptions to these guidelines. For now we plan to handle these on a case-by-case basis in the API reviews as we expect these will be uncommon.

> Azure services need to change over time. However, when changing a service, there are 2 requirements:

Reviewer 5: recommend comparing this with RP documentation. Data plane need not be the same as control plane, but good to know where we're different.

Author: +1. We have taken an action item to do this.

> Azure services need to change over time. However, when changing a service, there are 2 requirements:

Reviewer 5: agree with the don't break people principle. Beyond that - are we really saying there will never be any deprecations? What if there's a zero day exploit? Keep the API but handle through error codes? Would be good to point to a deprecation playbook, even if deprecation doesn't look like typical client API deprecation.

Reviewer 5: I see the info on this further down. I think it's good to be clear on the principles and not muddy them with exceptions however the exceptional case is rather a long way down the document such that i thought it hadn't been considered at all.

Author: +1 - We will link to the deprecation policy here. 
<sup>Note: This is an internal MS Doc.</sub>

Reviewer 3: I think this is addressed now -- see line 617.

### Distributed Tracing & Telemetry
Azure SDK client guidelines specify that client libraries must send telemetry data through the ```User-Agent``` header, ```X-MS-UserAgent``` header, and Open Telemetry. 
Client libraries are required to send telemetry and distributed tracing information on every  request. Telemetry information is vital to the effective operation of your service and should be a consideration from the outset of design and implementation efforts.

> :white_check_mark: **DO** follow the Azure SDK client guidelines for supporting telemetry headers and Open Telemetry.

> :no_entry: **DO NOT** reject a call if you have custom headers you don't understand, and specifically, distributed tracing headers. 
#### Additional References
* [Azure SDK client guidelines](https://azure.github.io/azure-sdk/general_azurecore.html)
* [Azure SDK User-Agent header policy](https://azure.github.io/azure-sdk/general_azurecore.html#azurecore-http-telemetry-x-ms-useragent)
* [Azure SDK Distributed tracing policy](https://azure.github.io/azure-sdk/general_azurecore.html#distributed-tracing-policy) 
* [Open Telemetry](https://opentelemetry.io/)

> ## Final thoughts

Reviewer 10: It's possible that I missed it, but I think we should have a section on granularity of service operations. From time to time we run into cases where a single user scenario corresponds to too many service calls. This means that client library has to to many calls to the service to actually do what the user cares about. This in turn is difficult to make reliable, explainable, etc. This issue could be mitigated by upleveling REST APIs or by providing batching APIs.

Author: I think there are two distinct thoughts in here. There first is the "upleveling" of an API. This is probably worth adding to the "Considerations" document. The second is batching, which feels like an new entry in the API patterns section. I'll add issues for both so we can review them with the larger team. 
See: #269 
See: #270

 | tenant | Globally-unique ID of container representing tenant isolation, billing, enforced quotas, lifetime of containees (ex: subscription UUID)
 | service&#x2011;root | Service-specific path (ex: blobcontainer, myqueue)
 | resource&#x2011;collection | Name of the collection, unabbreviated, pluralized
 | resource&#x2011;id | Value of the unique id property. This MUST be the raw string/number/guid value with no quoting but properly escaped to fit in a URL segment.

> :white_check_mark: **DO** use kebab-casing (preferred) or camel-casing for URL path segments. If the segment refers to a JSON field, use camel casing.
> 
> :white_check_mark: **DO** return ```414-URI Too Long``` if a URL exceeds 2083 characters
>
> :white_check_mark: **DO** treat URL path segments as case-sensitive. If the passed-in case doesn't match what the service expects, the request __MUST__ fail with the appropriate HTTP return code.

> :ballot_box_with_check: **YOU SHOULD** keep URLs readable; if possible, avoid UUIDs & %-encoding (ex: Cádiz is %-encoded as C%C3%A1diz)
> 

> > :ballot_box_with_check: **YOU SHOULD** limit your URL's characters to `0-9  A-Z  a-z  -  .  _  ~`

Reviewer 3: I think we need to add ":" to this list since we have decided that it should be used as the "Action" character.

Author: +1

> 
> :white_check_mark: **DO** compare request header values using case-sensitivity if the header name requires it
> 
> :white_check_mark: **DO** accept and return date values in headers using the HTTP Date format as defined in [RFC 7231, Section 7.1.1.1](https://datatracker.ietf.org/doc/html/rfc7231#section-7.1.1.1), e.g. "Sun, 06 Nov 1994 08:49:37 GMT"

> :no_entry: **DO NOT** fail a request that contains an unrecognized header. Headers may be added by API gateways or middleware and this must be tolerated
> 
> :no_entry: **DO NOT** use "x-" prefix for custom headers, unless the header already exists in production [[RFC 6648](https://datatracker.ietf.org/doc/html/rfc6648)].

#### Additional References
* [StackOverflow - Difference between http parameters and http headers](https://stackoverflow.com/questions/40492782)
* [Standard HTTP Headers](https://httpwg.org/specs/rfc7231.html#header.field.registration)
* [Why isn't HTTP PUT allowed to do partial updates in a REST API?](https://stackoverflow.com/questions/19732423/why-isnt-http-put-allowed-to-do-partial-updates-in-a-rest-api)

### REpresentational State Transfer (REST)

> REST is an architectural style with broad reach that emphasizes scalability, generality, independent deployment, reduced latency via caching, and security. When applying REST to your API, you define your service’s resources as a collections of items. These are typically the nouns you use in the vocabulary of your service. Your service's [URLs](#URLS) determine the hierarchical path developers use to perform CRUD (create, read, update, and delete) operations on resources. Note, it's important to model resource state, not behavior. There are patterns, later in these guidelines, that describe how to invoke behavior on your service.

Reviewer 11: Considering the controversy in the past about whether we were really RESTful in the Fielding sense, should we have some remarks about our definition of REST?

Reviewer 3: We could point to the Azure Architecture Center [description on REST](https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design#what-is-rest) which concludes with the statement:
> In practice, many published web APIs fall somewhere around level 2.

I don't have all the history on this, but I think the content there generally reflects the approach taken in these guidelines.

Author: +1

 Number  | Signed floating point (IEEE-754 binary64; int range: -2<sup>53</sup>+1 to +2<sup>53</sup>-1)
 String  | Used for everything else

> :white_check_mark: **DO** use integers within the acceptable range of JSON number.
> 
> :white_check_mark: **DO** establish a well-defined contract for the format of strings. For example, determine maximum length, legal characters, case-(in)sensitive comparisons, etc. Where possible, use standard formats, e.g. RFC3339 for date/time.
> 
> :white_check_mark: **DO** use strings formats that are well-known and easily parsable/formattable by many programming languages, e.g. RFC3339 for date/time.
> 
> :white_check_mark: **DO** ensure that information exchanged between your service and any client is "round-trippable" across multiple programming languages.
> 
> :white_check_mark: **DO** use [RFC3339] for date/time.
> 
> :white_check_mark: **DO** use [RFC4122] for UUIDs.

> :heavy_check_mark: **YOU MAY** use JSON objects to group sub-fields together.
> 
> :heavy_check_mark: **YOU MAY** use JSON arrays if maintaining an order of values is required.

> :ballot_box_with_check: **YOU SHOULD** use JSON objects instead of arrays whenever possible.

> > :warning: **YOU SHOULD NOT** use JSON Arrays, e.g. [ value, … ]. Arrays are very difficult and inefficient to work with, especially with ```JSON Merge Patch```, as the entire array needs to be read prior to any operation being applied to it.

Reviewer 11: I would move this closer to the MAY use JSON arrays and clarify the conflict between MAY and SHOULD NOT

Author: Got it--I can see the confusion here & will clarify.

 
> :no_entry: **DO NOT** send "enum integers" over the wire.
> 
> :no_entry: **DO NOT** remove values from your enumeration list as this breaks customer code.

#### Polymorphic types
> :warning: **YOU SHOULD NOT** use polymorphic JSON types because they greatly complicate the customer code due to runtime dynamic casts and the introduction of new types in the future. 

If you can't avoid them, then follow the guideline below.

> :white_check_mark: **DO** define a ```kind``` field indicating the kind of the resource and include any kind-specific fields in the body. 

Below is an example of JSON for a Rectangle and Circle:
**Rectangle**
```json
{
   "kind": "rectangle", 
   "x": 100, 
   "y": 50, 
   "width": 10, 
   "length": 24, 
   "fillColor": "Red", 
   "lineColor": "White", 
   "subscription": {
      "kind": "free"
   }
}
```

**Circle**
 ```json
{
   "kind": "circle",
   "x": 100, 
   "y": 50, 
   "radius": 10, 
   "fillColor": "Green", 
   "lineColor": "Black", 
   "subscription": { 
      "kind": "paid", 
      "expiration": "2024", 
      "invoice": "123456"
   }
}
```

> Both Rectangle and Circle has common fields: ```kind```, ```fillColor```, ```lineColor```, and ```subscription```. A Rectangle also has ```x```, ```y```, ```width```, and ```length``` while a Circle has ```x```, ```y```, and ```radius```. The ```subscription``` is a nested polymorphic type. A ```free``` subscription has no additional fields and a ```paid``` subscription has ```expiration``` and ```invoice``` fields.

Reviewer 11: *have

> :no_entry: **DO NOT** return the ```nextLink``` field at all when returning the last page of the collection. 
>
> :no_entry: **DO NOT** ever return a ```nextLink``` field with a value of null.

#### Query options
> :heavy_check_mark: **YOU MAY** support the following query parameters allowing customers to control the list operation:

Parameter&nbsp;name | Type | Description
------------------- | ---- | -----------
_filter_       | string            | an expression on the resource type that selects the resources to be returned
_orderby_      | string&nbsp;array | a list of expressions that specify the order of the returned resources
_skip_         | integer           | an offset into the collection of the first resource to be returned
_top_          | integer           | the maximum number of resources to return from the collection
_maxpagesize_  | integer           | the maximum number of resources to include in a single response
_select_       | string&nbsp;array | a list of field names to be returned for each resource
_expand_       | string&nbsp;array | a list of the related resources to be included in line with each resource

> 

Reviewer 2: NOTE: The orderby parameter can be extremely expensive for a service to implement as it must retrieve all items to sort them. And if the operation supports paging (which is likely), then a client request to get another page may have to retrieve all items and sort them again to determine which items are on the desired page. Only support orderby if customer scenarios really demand it and the service is confident that it can support it in perpetuity (even if the backing storage service changes someday).

> :white_check_mark: **DO** respond with an error message as defined in the [Handling Errors](handling-errors) section if a client includes an operator in a _filter_ expression that is not supported by the operation.
> 
> :white_check_mark: **DO** use the following operator precedence for supported operators when evaluating _filter_ expressions. Operators are listed by category in order of precedence from highest to lowest. Operators in the same category have equal precedence and should be evaluated left to right:

| Group           | Operator | Description
| ----------------|----------|------------
| Grouping        | ( )      | Precedence grouping   |
| Unary           | not      | Logical Negation      |
| Relational      | gt       | Greater Than          |
|                 | ge       | Greater than or Equal |
|                 | lt       | Less Than             |
|                 | le       | Less than or Equal    |
| Equality        | eq       | Equal                 |
|                 | ne       | Not Equal             |
| Conditional AND | and      | Logical And           |
| Conditional OR  | or       | Logical Or            |

> 

Reviewer 2: > :heavy_check_mark: **YOU MAY** support orderby and filter functions such as concat and contains. For more information, see [odata Canonical Functions](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#_Toc31360979).

> 
> :white_check_mark: **DO** return the collection's _top_ number of resources (if available), starting from _skip_.

##### maxpagesize

> :heavy_check_mark: **YOU MAY** allow clients to pass the _maxpagesize_ query parameter to specify the maximum number of resources to include in a single page response.

If supporting _maxpagesize_

> :white_check_mark: **DO** define the _maxpagesize_ parameter as an optional integer with a default value appropriate for the collection.
> 
> :white_check_mark: **DO** make clear in documentation of the _maxpagesize_ parameter that the operation may choose to return fewer resources than the value specified.

### API Versioning

Azure services need to change over time. However, when changing a service, there are 2 requirements:
 1. Already-running customer workloads must never break due to a service change
 2. Customers can adopt a new service version without requiring any code changes (Of course, the customer must modify code to leverage any new service features.)

> :white_check_mark: **DO** review any API changes with the Azure API Stewardship Board
> 

> > :white_check_mark: **DO** use an 'api-version' query parameter with a date value

Reviewer 3: The parameter name `api-version` contradicts our own naming guidance for query parameters: lower camel case.

Reviewer 8: very good point, yes, but its thoroughly established at this point. also helps keep "cross-cutting" parameters separate from service specific ones

> ```text PUT https://service.azure.com/users/Jeff?api-version=2021-06-04```
>
> :white_check_mark: **DO** use a later date for each new preview version
> When releasing a new preview, the service team may completely retire any previous preview versions after giving customers at least 90 days to upgrade their code

> :white_check_mark: **DO** use a later date for successive preview versions. 

> :no_entry: **DO NOT** use the same date when transitioning from a preview API to a GA API. If the preview 'api-version' is '2021-06-04-preview', the GA version of the API <b>must be</b> a date later than 2021-06-04
>
> :no_entry: **DO NOT** keep a preview feature in preview for more than 1 year; it must go GA (or be removed) within 1 year after introduction.
> 
> :no_entry: **DO NOT** introduce any breaking changes into the service. 
*NOTE: the [Azure Breaking Change Policy](http://aka.ms/AzBreakingChangesPolicy/) has tables (section 5) describing what kinds of changes are considered breaking. If a new service version must break customers (due to security/compliance/etc.), contact the [Azure Breaking Change Reviewers](mailto:azbreakchangereview@microsoft.com) as soon as possible.*

> 

Reviewer 3: I think we should add:
```
> :no_entry: **DO NOT** include a version number segment in any operation paths.
```

Reviewer 3: Added above.

Author: LGTM.

 | tenant | Globally-unique ID of container representing tenant isolation, billing, enforced quotas, lifetime of containees (ex: subscription UUID)
 | service&#x2011;root | Service-specific path (ex: blobcontainer, myqueue)
 | resource&#x2011;collection | Name of the collection, unabbreviated, pluralized
 | resource&#x2011;id | Value of the unique id property. This MUST be the raw string/number/guid value with no quoting but properly escaped to fit in a URL segment.

> :white_check_mark: **DO** use kebab-casing (preferred) or camel-casing for URL path segments. If the segment refers to a JSON field, use camel casing.
> 
> :white_check_mark: **DO** return ```414-URI Too Long``` if a URL exceeds 2083 characters
>
> :white_check_mark: **DO** treat URL path segments as case-sensitive. If the passed-in case doesn't match what the service expects, the request __MUST__ fail with a ```404-Not found``` HTTP return code.

> :ballot_box_with_check: **YOU SHOULD** keep URLs readable; if possible, avoid UUIDs & %-encoding (ex: Cádiz is %-encoded as C%C3%A1diz)
> 

> > :ballot_box_with_check: **YOU SHOULD** limit your URL's characters to `0-9  A-Z  a-z  -  .  _  ~ :`

Reviewer 7: If we are recommending either kebab-case or camelCase for path segments (see line 69), then I think we should move the “_” character from the “YOU SHOULD” to the “YOU MAY” list of URL characters.

Reviewer 2: The characters that never need %-encoding are 0-9  A-Z  a-z  -  .  _  ~; not colon. So I think colon should be removed from the list. 

If you want: there is a difference between service-defined path segments and customer-specified path segments. I think the service should stick with 0-9  A-Z  a-z  -  .  _  ~ (so make this a DO guideline; not YOU SHOULD). We reserve colon for actions only (so don't list it here so that service teams avoid it). 

Customers can use 0-9  A-Z  a-z  -  .  _  ~ and other %-encoded values. They should avoid colon and other special characters. For example, what if the customer uses quote? Then we might return a json field with "id":"database"Name" (quote between database & name on purpose) - this is not valid JSON.
Or, if the customer uses ?, then the path ends and the query parameters begin - bad.
Or, if they use /, then they are inserting a path segment - bad.
Or, if they use #, then they are starting a fragment - bad.
Or, if they use +, then sometimes its interpreted as a space - bad.

Service teams don't think about any of this so it would be good to be more explicit in these guidelines for service vs customer characters. 
Customers segments MAY support the following characters (or a subset thereof): 0-9  A-Z  a-z  -  .  _  ~  !  $  &   '  (  )  *  ,   ;  =  @  [] (space). NOTE: I took a stab at this list but am open to additions/removals.

#### Exactly Once Behavior = Client Retries & Service Idempotency

:white_check_mark: **DO** ensure that _all_ HTTP methods are idempotent.

:ballot_box_with_check: **YOU SHOULD** use PUT or PATCH to create a resource as these HTTP methods are easy to implement, allow the customer to name their own resource, and are idempotent.

:heavy_check_mark: **YOU MAY** use POST to create a resource but you must make it idempotent and, of course, the response **MUST** return the URL of the created resource with a 201-Created. One way to make POST idempotent is to use the Repeatability-Request-ID & Repeatability-First-Sent headers (See [Repeatability of requests](#Repeatability-of-requests)).

##### HTTP Return Codes

:white_check_mark: **DO** adhere to the return codes in the following table when the method completes synchronously and is successful:

Method | Description | Response Status Code
-------|-------------|---------------------
PATCH  | Create/Modify the resource with JSON Merge Patch | `200-OK`, `201-Created`
PUT    | Create/Replace the _whole_ resource | `200-OK`, `201-Created`

> POST   | Create new resource (ID set by service) | `201-Created` with URL of created resource

Reviewer 7: I think we probably want to allow `200-OK` on POSTs as well, right?

Reviewer 2: Maybe. If it is idempotent (due to the special Oasis header), then 200 could be returned if the operation is being retried. But, in this case, the client still think it is creating a new resource and then 201 let's it know that th resource was created. So, I think we can stick with 201 only.

Reviewer 3: But what about POST when used for action operations, like "reboot VM" or "translate this string"?  Those should return `200-OK`, right?

Reviewer 2: Yes, but I think that should be called out specifically in the action section; not in this section.
This section is about resource CRUD operations ONLY.

Reviewer 3: Added to the "Performing an Action" section.

--------- | -------
Boolean   | true / false (all lowercase)
Integer   | -2<sup>53</sup>+1 to +2<sup>53</sup>-1 (for consistency with JSON limits on integers [RFC8259](https://datatracker.ietf.org/doc/html/rfc8259))
Float     | [IEEE-754 binary64](https://en.wikipedia.org/wiki/Double-precision_floating-point_format)
String    | (Un)quoted?, max length, legal characters, case-sensitive, multiple delimiter
UUID      | 123e4567-e89b-12d3-a456-426614174000 (no {}s, hyphens, case-insensitive) [RFC4122](https://datatracker.ietf.org/doc/html/rfc4122)
Date/Time (Header) | Sun, 06 Nov 1994 08:49:37 GMT [RFC7231](https://datatracker.ietf.org/doc/html/rfc7231#section-7.1.1.1)
Date/Time (Query parameter) | YYYY-MM-DDTHH:mm:ss.sssZ (with at most 3 digits of fractional seconds) [RFC3339](https://datatracker.ietf.org/doc/html/rfc3339)
Byte array | Base-64 encoded, max length
Array      | One of a) a comma-separated list of values (preferred), or b) separate `name=value` parameter instances for each value of the array

The table below lists the headers most used by Azure services:

Header Key          | Applies to | Example
------------------- | ---------- | -------------

> _authorization_     | Request    | Bearer eyJ0...Xd6j (Support Azure Active Directory)

Reviewer 3: Should "Azure Active Directory" be a link to the information on how to do this?

Reviewer 2: Ideally yes. I don't have that link thought. Scott Schaab might have it.
